braindecode/preprocessing/preprocess.py

```diff
@@ -120,54 +127,109 @@ def preprocess(concat_ds, preprocessors):
+    #     is provided.
+    setattr(concat_ds, concat_kind + '_preproc_kwargs', preproc_kwargs)
+
+    return concat_ds
 
-    for ds in concat_ds.datasets:
-        if hasattr(ds, 'raw'):
-            _preprocess(ds.raw, preprocessors)
-            # store raw preprocessing keyword arguments to the dataset
-            setattr(concat_ds, 'raw_preproc_kwargs', preproc_kwargs)
-        elif hasattr(ds, 'windows'):
-            _preprocess(ds.windows, preprocessors)
-            # store window preprocessing keyword arguments to the dataset
-            setattr(concat_ds, 'window_preproc_kwargs', preproc_kwargs)
-        else:
-            raise ValueError(
-                'Can only preprocess concatenation of BaseDataset or '
-                'WindowsDataset, with either a `raw` or `windows` attribute.')
 
-    # Recompute cumulative sizes as the transforms might have changed them
-    # XXX: Ultimately, the best solution would be to have cumulative_size be
-    #      a property of BaseConcatDataset.
-    concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+def _replace_inplace(concat_ds, new_concat_ds):
+    if len(concat_ds.datasets) != len(new_concat_ds.datasets):
+        raise ValueError('Both inputs must have the same length.')
+    for i in range(len(new_concat_ds.datasets)):
+        concat_ds.datasets[i] = new_concat_ds.datasets[i]
 
 
-def _preprocess(raw_or_epochs, preprocessors):
+def _preprocess(ds, ds_index, preprocessors, save_dir=None, overwrite=False):
     """Apply preprocessor(s) to Raw or Epochs object.
 
     Parameters
     ----------
-    raw_or_epochs: mne.io.Raw or mne.Epochs
-        Object to preprocess.
+    ds: BaseDataset | WindowsDataset
+        Dataset object to preprocess.
+    ds_index : int
+        Index of the BaseDataset in its BaseConcatDataset. Ignored if save_dir
+        is None.
     preprocessors: list(Preprocessor)
         List of preprocessors to apply to the dataset.
+    save_dir : str | None
+        If provided, save the preprocessed BaseDataset in the
+        specified directory.
+    overwrite : bool
+        If True, overwrite existing file with the same name.
     """
-    for preproc in preprocessors:
-        preproc.apply(raw_or_epochs)
+
+    def _preprocess_raw_or_epochs(raw_or_epochs, preprocessors):
+        for preproc in preprocessors:
+            preproc.apply(raw_or_epochs)
+
+    if hasattr(ds, 'raw'):
+        _preprocess_raw_or_epochs(ds.raw, preprocessors)
+    elif hasattr(ds, 'windows'):
+        _preprocess_raw_or_epochs(ds.windows, preprocessors)
+    else:
+        raise ValueError(
+            'Can only preprocess concatenation of BaseDataset or '
+            'WindowsDataset, with either a `raw` or `windows` attribute.')
+
+    if save_dir is not None:
+        concat_ds = BaseConcatDataset([ds])
+        save_subdir = os.path.join(save_dir, str(ds_index))
+        # XXX Save target_name.json if ds is a BaseDataset
```

Comment:
Should be done automatically when calling `BaseConcatDataset.save()`, see https://github.com/braindecode/braindecode/blob/master/braindecode/datasets/base.py#L419-L420.